compiler/lib/driver.ml

```diff
@@ -677,31 +691,39 @@ let full ~standalone ~wrap_with_fun ~profile ~link ~source_map formatter d p =
        | O3 -> o3)
     +> exact_calls ~deadcode_sentinal profile
     +> effects ~deadcode_sentinal
-    +> map_fst (if Config.Flag.effects () then fun x -> x else Generate_closure.f)
+    +> map_fst
+         (match Config.target () with
+         | `JavaScript -> if Config.Flag.effects () then Fun.id else Generate_closure.f
+         | `Wasm -> Fun.id)
     +> map_fst deadcode'
   in
-  let emit =
+  if times () then Format.eprintf "Start Optimizing...@.";
+  let t = Timer.make () in
+  let (program, variable_uses), trampolined_calls, in_cps = opt p in
+  let () = if times () then Format.eprintf " optimizations : %a@." Timer.print t in
+  { program; variable_uses; trampolined_calls; in_cps; deadcode_sentinal }
+
+let full ~standalone ~wrap_with_fun ~profile ~link ~source_map ~formatter d p =
+  let { program; variable_uses; trampolined_calls; in_cps; deadcode_sentinal } =
+    optimize ~profile p
+  in
+  let exported_runtime = not standalone in
+  let emit formatter =
     generate
       d
       ~exported_runtime
       ~wrap_with_fun
       ~warn_on_unhandled_effect:standalone
       ~deadcode_sentinal
-    +> link' ~export_runtime ~standalone ~link
-    +> pack ~wrap_with_fun ~standalone
-    +> coloring
-    +> check_js
+    +> link_and_pack ~standalone ~wrap_with_fun ~link
     +> output formatter ~source_map ()
   in
-  if times () then Format.eprintf "Start Optimizing...@.";
-  let t = Timer.make () in
-  let r = opt p in
-  let () = if times () then Format.eprintf " optimizations : %a@." Timer.print t in
-  emit r
+  let source_map = emit formatter ((program, variable_uses), trampolined_calls, in_cps) in
```

Comment:
The nested tuples doesn't make sense anymore, maybe one should just pass an optimized result.

Reply:
It’s indeed better. Done.